File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.js
function solution(storage, requests) {
  let N = storage.length;
  let M = storage[0].length;
  let answer = 0;

  const updated = Array(N + 2).fill(0).map((_, i) => {
    if (i === 0 || i === N + 1) return Array(M + 2).fill('0');
    return ['0', ...storage[i - 1].split(''), '0'];
  });

  const direction = [
    [1, 0], [0, 1], [-1, 0], [0, -1]
  ];

  function listForklift(command) {
    // front-rear 기반 큐 구현
    const MAX = (N + 2) * (M + 2);
    const queue = new Array(MAX);
    let front = 0, rear = 0;

    const visited = Array.from({ length: N + 2 }, () => Array(M + 2).fill(false));

    queue[rear++] = [0, 0];
    visited[0][0] = true;

    while (front < rear) {
      const [cx, cy] = queue[front++];

      for (let [dx, dy] of direction) {
        const nx = cx + dx;
        const ny = cy + dy;

        if (nx >= 0 && nx < N + 2 && ny >= 0 && ny < M + 2 && !visited[nx][ny]) {
          const cell = updated[nx][ny];

          if (cell === '0') {
            queue[rear++] = [nx, ny]; // 통과만
          } else if (cell === command) {
            updated[nx][ny] = '0'; // 꺼내기
          }

          visited[nx][ny] = true;
        }
      }
    }
  }

  function listCrain(command) {
    const target = command[0];

    for (let i = 0; i < N + 2; i++) {
      for (let j = 0; j < M + 2; j++) {
        if (updated[i][j] === target) {
          updated[i][j] = '0';
        }
      }
    }
  }

  for (const command of requests) {
    if (command.length === 1) {
      listForklift(command);
    } else {
      listCrain(command);
    }
  }

  for (let i = 0; i < N + 2; i++) {
    for (let j = 0; j < M + 2; j++) {
      if (updated[i][j] !== '0') {
        answer++;
      }
    }
  }

  return answer;
}

File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.py
from collections import deque

def solution(storage, requests):
    answer = 0
    N, M = len(storage), len(storage[0])
    updated = [['0'] * (M+2)];
    updated.extend([['0']+list(str)+['0'] for str in storage]);
    updated.append(['0'] * (M+2));
    
    direction = [[1,0],[0,1],[-1,0],[0,-1]]
    
    def list_forklift(command):
        # print("지게차", command)
        #BFS
        def BFS (x, y):
            nonlocal updated
            visited = [[False]*(M+2) for i in range(N+2)]
            
            queue = deque()
            queue.append((x, y))
            visited[x][y] = True
            
            while(len(queue) > 0):
                cx, cy = queue.popleft()
                for dx, dy in direction:
                    nx,ny = cx+dx, cy+dy;
                    if(0<=nx<N+2 and 0<=ny<M+2 and not visited[nx][ny]):
                        # print("way", cx, cy ,"=>", nx, ny)
                        if(updated[nx][ny] == '0'):
                            queue.append((nx, ny))
                        elif(updated[nx][ny] == command):
                            updated[nx][ny] = '0' # 꺼내기
                        visited[nx][ny] = True
                            # print("Right way", cx, cy ,"=>", nx, ny)
                    
        BFS(0,0);
            
        return 1
        
    def list_crain(command):
        nonlocal updated
        temp = []
        updated = [
            ['0' if c ==command[0] else c for c in row]
            for row in updated
        ]
        # print("크레인", command[:1])
        return 2
    
    for command in requests:
        if(len(command) == 1):
            list_forklift(command)
        else :
            list_crain(command)
            
        # print(updated)

    for row in updated:
        for str in row:
            if(not str == '0'):
                answer += 1
    return answer